File: tests/test_matching.py
"""Matching semantics: price-time priority, sweeps, partial fills, market orders."""

from __future__ import annotations

import random

import pytest

from orderbook import (
    BookIntegrityError,
    MatchingEngine,
    Order,
    OrderBook,
    OrderStatus,
    OrderType,
    Side,
)


# ----------------------------------------------------------------------
# No match
# ----------------------------------------------------------------------


def test_non_crossing_limit_rests(market):
    market.sell(10, 101.0)
    result = market.buy(10, 100.0)

    assert result.trades == []
    assert result.resting
    assert result.status is OrderStatus.RESTING
    assert market.book.best_bid() == 100.0
    market.validate()


def test_limit_one_tick_away_does_not_trade(market):
    market.sell(10, 100.01)
    result = market.buy(10, 100.0)
    assert result.trades == []
    market.validate()


def test_first_order_into_an_empty_book_rests(market):
    result = market.buy(10, 100.0)
    assert result.trades == []
    assert result.resting


# ----------------------------------------------------------------------
# Basic execution
# ----------------------------------------------------------------------


def test_exact_match_clears_both_orders(market):
    resting = market.sell(10, 100.0).order
    result = market.buy(10, 100.0)

    assert len(result.trades) == 1
    trade = result.trades[0]
    assert trade.price == 100.0
    assert trade.quantity == 10
    assert trade.aggressor_side is Side.BUY
    assert trade.resting_order_id == resting.order_id

    assert result.status is OrderStatus.FILLED
    assert not result.resting
    assert resting.status is OrderStatus.FILLED
    assert len(market.book) == 0
    assert market.book.best_ask() is None
    market.validate()


def test_trade_prints_at_the_resting_price(market):
    """An aggressive buy gets price improvement, not its own limit."""
    market.sell(10, 100.0)
    result = market.buy(10, 105.0)

    assert result.trades[0].price == 100.0
    assert result.average_price == pytest.approx(100.0)


def test_aggressor_side_is_recorded(market):
    market.buy(10, 100.0)
    result = market.sell(10, 100.0)
    assert result.trades[0].aggressor_side is Side.SELL


def test_trade_buy_sell_ids_follow_the_aggressor(market):
    resting_ask = market.sell(5, 100.0).order
    aggressive_bid = market.buy(5, 100.0)
    trade = aggressive_bid.trades[0]
    assert trade.buy_order_id == aggressive_bid.order.order_id
    assert trade.sell_order_id == resting_ask.order_id


def test_trade_ids_are_sequential(market):
    market.sell(1, 100.0)
    market.sell(1, 101.0)
    result = market.buy(2, 101.0)
    assert [t.trade_id for t in result.trades] == [1, 2]


# ----------------------------------------------------------------------
# Partial fills
# ----------------------------------------------------------------------


def test_incoming_larger_than_resting_rests_the_remainder(market):
    market.sell(4, 100.0)
    result = market.buy(10, 100.0)

    assert result.filled_quantity == 4
    assert result.order.remaining == 6
    assert result.resting
    assert result.status is OrderStatus.PARTIALLY_FILLED
    assert market.book.best_bid() == 100.0
    assert market.book.bid_depth == 6
    market.validate()


def test_resting_larger_than_incoming_keeps_its_place(market):
    resting = market.sell(10, 100.0).order
    market.buy(4, 100.0)

    assert resting.remaining == 6
    assert resting.status is OrderStatus.PARTIALLY_FILLED
    assert market.book.best_ask() == 100.0
    assert market.book.ask_depth == 6
    assert market.book.l2().asks == [(100.0, 6, 1)]
    market.validate()


def test_partial_fill_does_not_lose_time_priority(market):
    first = market.sell(10, 100.0).order
    second = market.sell(10, 100.0).order

    market.buy(4, 100.0)  # eats into `first` only

    assert [o.order_id for o in market.book.orders(Side.SELL)] == [
        first.order_id,
        second.order_id,
    ]


# ----------------------------------------------------------------------
# Price-time priority
# ----------------------------------------------------------------------


def test_best_price_fills_before_worse_price(market):
    worse = market.sell(5, 101.0).order
    better = market.sell(5, 100.0).order

    result = market.buy(5, 101.0)

    assert result.trades[0].resting_order_id == better.order_id
    assert market.book.contains(worse.order_id)
    market.validate()


def test_fifo_within_a_price_level(market):
    first = market.sell(5, 100.0).order
    second = market.sell(5, 100.0).order
    third = market.sell(5, 100.0).order

    result = market.buy(15, 100.0)
    hit = [t.resting_order_id for t in result.trades]
    assert hit == [first.order_id, second.order_id, third.order_id]
    market.validate()


def test_sweep_walks_levels_in_price_order(market):
    market.sell(5, 100.0)
    market.sell(5, 101.0)
    market.sell(5, 102.0)

    result = market.buy(15, 102.0)

    assert [t.price for t in result.trades] == [100.0, 101.0, 102.0]
    assert result.filled_quantity == 15
    assert result.traded_notional == pytest.approx(5 * (100.0 + 101.0 + 102.0))
    assert result.average_price == pytest.approx(101.0)
    assert market.book.best_ask() is None
    market.validate()


def test_sweep_stops_at_the_limit_price(market):
    market.sell(5, 100.0)
    market.sell(5, 101.0)
    market.sell(5, 102.0)

    result = market.buy(15, 101.0)

    assert result.filled_quantity == 10
    assert result.order.remaining == 5
    assert result.resting
    assert market.book.best_ask() == 102.0
    assert market.book.best_bid() == 101.0
    market.validate()


def test_sell_sweep_walks_bids_high_to_low(market):
    market.buy(5, 100.0)
    market.buy(5, 99.0)
    market.buy(5, 98.0)

    result = market.sell(15, 98.0)
    assert [t.price for t in result.trades] == [100.0, 99.0, 98.0]
    market.validate()


# ----------------------------------------------------------------------
# Market orders
# ----------------------------------------------------------------------


def test_market_buy_sweeps_regardless_of_price(market):
    market.sell(5, 100.0)
    market.sell(5, 500.0)

    result = market.market(Side.BUY, 10)

    assert [t.price for t in result.trades] == [100.0, 500.0]
    assert result.status is OrderStatus.FILLED
    market.validate()


def test_market_order_never_rests(market):
    market.sell(4, 100.0)
    result = market.market(Side.BUY, 10)

    assert result.filled_quantity == 4
    assert result.order.remaining == 6
    assert not result.resting
    assert result.status is OrderStatus.EXPIRED
    assert market.book.best_bid() is None
    assert len(market.book) == 0
    market.validate()


def test_market_order_against_empty_book_expires(market):
    result = market.market(Side.BUY, 10)
    assert result.trades == []
    assert result.status is OrderStatus.EXPIRED
    assert len(market.book) == 0


def test_market_sell_hits_the_bid(market):
    market.buy(10, 99.0)
    result = market.market(Side.SELL, 10)
    assert result.trades[0].price == 99.0
    assert result.status is OrderStatus.FILLED
    market.validate()


# ----------------------------------------------------------------------
# Empty price levels
# ----------------------------------------------------------------------


def test_level_is_collected_when_fully_traded(market):
    market.sell(5, 100.0)
    assert market.book.level_count(Side.SELL) == 1

    market.buy(5, 100.0)

    assert market.book.level_count(Side.SELL) == 0
    assert market.book.level_removes == 1
    assert market.book.l2().asks == []
    market.validate()


def test_partially_traded_level_is_kept(market):
    market.sell(10, 100.0)
    market.buy(4, 100.0)

    assert market.book.level_count(Side.SELL) == 1
    assert market.book.level_removes == 0
    assert market.book.l2().asks == [(100.0, 6, 1)]
    market.validate()


def test_level_with_a_second_order_survives_a_full_fill(market):
    first = market.sell(5, 100.0).order
    second = market.sell(5, 100.0).order

    market.buy(5, 100.0)  # consumes `first` entirely

    assert not market.book.contains(first.order_id)
    assert market.book.contains(second.order_id)
    assert market.book.level_count(Side.SELL) == 1
    assert market.book.level_removes == 0
    market.validate()


def test_sweep_collects_every_level_it_empties(market):
    for price in (100.0, 101.0, 102.0):
        market.sell(5, price)
    assert market.book.level_count(Side.SELL) == 3

    market.buy(15, 102.0)

    assert market.book.level_count(Side.SELL) == 0
    assert market.book.level_removes == 3
    assert market.book.ask_depth == 0
    market.validate()


def test_price_is_reusable_after_being_traded_out(market):
    market.sell(5, 100.0)
    market.buy(5, 100.0)
    assert market.book.best_ask() is None

    market.sell(3, 100.0)

    assert market.book.best_ask() == 100.0
    assert market.book.l2().asks == [(100.0, 3, 1)]
    market.validate()


def test_no_empty_level_is_ever_left_behind(market):
    """Sweep the whole book away and assert nothing phantom remains."""
    for price in (99.0, 99.5, 100.0):
        market.buy(4, price)
    for price in (101.0, 101.5, 102.0):
        market.sell(4, price)

    market.market(Side.BUY, 12)
    market.market(Side.SELL, 12)

    assert len(market.book) == 0
    assert market.book.level_count(Side.BUY) == 0
    assert market.book.level_count(Side.SELL) == 0
    assert market.book.l2().bids == []
    assert market.book.l2().asks == []
    assert market.book.level_creates == market.book.level_removes
    market.validate()


# ----------------------------------------------------------------------
# Crossed and locked books
# ----------------------------------------------------------------------


def test_a_crossing_limit_trades_instead_of_resting(market):
    market.sell(5, 100.0)
    result = market.buy(5, 101.0)

    assert result.trades
    assert not result.resting
    assert market.book.best_bid() is None
    market.validate()


def test_a_locked_price_trades_rather_than_locking(market):
    """A bid equal to the best ask must trade — the book may not sit locked."""
    market.sell(5, 100.0)
    result = market.buy(5, 100.0)

    assert len(result.trades) == 1
    assert market.book.best_bid() is None
    assert market.book.best_ask() is None
    market.validate()


def test_book_is_uncrossed_after_a_deep_sweep(market):
    for price in (100.0, 100.5, 101.0, 101.5):
        market.sell(5, price)
    for price in (99.5, 99.0):
        market.buy(5, price)

    market.buy(14, 101.5)

    best_bid, best_ask = market.book.best_bid(), market.book.best_ask()
    assert best_bid is not None and best_ask is not None
    assert best_bid < best_ask
    market.validate()


def test_remainder_rests_without_crossing_the_far_side(market):
    """A partially filled aggressor rests at its own price, never through the book."""
    market.sell(5, 100.0)
    market.sell(5, 103.0)

    result = market.buy(10, 101.0)

    assert result.filled_quantity == 5
    assert market.book.best_bid() == 101.0
    assert market.book.best_ask() == 103.0
    assert market.book.best_bid() < market.book.best_ask()
    market.validate()


def test_repeated_crossing_flow_never_leaves_a_crossed_book(market):
    """Alternate aggressive both ways; the book must be uncrossed at every step."""
    for i in range(60):
        if i % 2:
            market.buy(7, 100.0 + (i % 5) * 0.01)
        else:
            market.sell(7, 100.0 - (i % 5) * 0.01)

        best_bid, best_ask = market.book.best_bid(), market.book.best_ask()
        if best_bid is not None and best_ask is not None:
            assert best_bid < best_ask, f"crossed at step {i}"
        market.validate()


def test_validate_detects_a_forced_crossed_book(book):
    """Bypass matching to plant a crossed state; validate must reject it."""
    book.add(Order(1, Side.BUY, OrderType.LIMIT, 5, 101.0))
    book.add(Order(2, Side.SELL, OrderType.LIMIT, 5, 100.0))

    with pytest.raises(BookIntegrityError, match="crossed book"):
        book.validate()


# ----------------------------------------------------------------------
# Conservation and invariants
# ----------------------------------------------------------------------


def test_book_never_stays_crossed(market):
    market.sell(10, 100.0)
    market.buy(10, 105.0)
    market.validate()  # raises if best bid >= best ask


def test_engine_trade_counter_matches_trades_emitted(market):
    market.sell(5, 100.0)
    market.sell(5, 101.0)
    market.buy(10, 101.0)
    assert market.engine.trades_generated == 2


def test_quantity_is_conserved_under_random_flow():
    """Traded volume + resting depth must equal everything ever submitted."""
    rng = random.Random(42)
    book = OrderBook()
    engine = MatchingEngine()

    submitted_buy = submitted_sell = 0
    bought = sold = 0

    for order_id in range(1, 2001):
        side = rng.choice([Side.BUY, Side.SELL])
        quantity = rng.randint(1, 25)
        price = round(rng.uniform(98.0, 102.0), 2)
        result = engine.submit(
            Order(order_id, side, OrderType.LIMIT, quantity, price), book
        )

        if side is Side.BUY:
            submitted_buy += quantity
        else:
            submitted_sell += quantity

        traded = sum(t.quantity for t in result.trades)
        bought += traded if side is Side.BUY else 0
        sold += traded if side is Side.SELL else 0
        # Each trade also fills the opposing resting order.
        if side is Side.BUY:
            sold += traded
        else:
            bought += traded

    book.validate()
    assert bought == sold  # every share bought was sold
    assert submitted_buy - bought == book.bid_depth
    assert submitted_sell - sold == book.ask_depth


def test_matching_is_deterministic_under_a_fixed_seed():
    def run():
        rng = random.Random(7)
        book = OrderBook()
        engine = MatchingEngine()
        prints = []
        for order_id in range(1, 501):
            side = rng.choice([Side.BUY, Side.SELL])
            order = Order(
                order_id,
                side,
                OrderType.LIMIT,
                rng.randint(1, 15),
                round(rng.uniform(99.0, 101.0), 2),
            )
            for trade in engine.submit(order, book).trades:
                prints.append((trade.price, trade.quantity, trade.resting_order_id))
        return prints, book.best_bid(), book.best_ask()

    assert run() == run()

File: cpp/tests/test_matching.cpp
/// Matching behavior: sweeps, locked/crossed books, market orders, views,
/// and one conservation property over randomized limit flow.

#include <random>
#include <vector>

#include "orderbook/matching_engine.hpp"
#include "check.h"

using namespace orderbook;

namespace {

bool multi_level_sweep_walks_in_price_order() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 101.0), book);
    engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(3, Side::SELL, OrderType::LIMIT, 5, 102.0), book);

    MatchResult r = engine.submit(Order(4, Side::BUY, OrderType::LIMIT, 12, 102.0), book);
    CHECK(r.trades.size() == 3);
    CHECK(r.trades[0].price == 100.0 && r.trades[0].quantity == 5);
    CHECK(r.trades[0].resting_order_id == 2);
    CHECK(r.trades[1].price == 101.0 && r.trades[1].quantity == 5);
    CHECK(r.trades[1].resting_order_id == 1);
    CHECK(r.trades[2].price == 102.0 && r.trades[2].quantity == 2);  // partial at the end
    CHECK(r.trades[2].resting_order_id == 3);
    CHECK(r.order.is_filled());
    CHECK(r.status() == OrderStatus::FILLED);
    CHECK(book.ask_depth() == 3);
    book.validate();
    return true;
}

bool locked_book_trades_at_the_touch() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);

    // Incoming bid at exactly the ask: a locked print, executed at 100.
    MatchResult r = engine.submit(Order(2, Side::BUY, OrderType::LIMIT, 5, 100.0), book);
    CHECK(r.trades.size() == 1);
    CHECK(r.trades[0].price == 100.0);
    CHECK(r.order.is_filled());
    CHECK(book.size() == 0);
    book.validate();
    return true;
}

bool crossed_incoming_limit_gets_improvement_then_rests_uncrossed() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 5, 101.0), book);
    engine.submit(Order(3, Side::SELL, OrderType::LIMIT, 5, 103.0), book);

    // Willing to pay 102: sweeps 100 and 101 (price improvement), then the
    // remainder rests at 102 — inside the spread, never crossing 103.
    MatchResult r = engine.submit(Order(4, Side::BUY, OrderType::LIMIT, 20, 102.0), book);
    CHECK(r.trades.size() == 2);
    CHECK(r.trades[0].price == 100.0);
    CHECK(r.trades[1].price == 101.0);
    CHECK(r.resting);
    CHECK(r.order.remaining == 10);
    CHECK(r.status() == OrderStatus::PARTIALLY_FILLED);
    CHECK(book.best_bid() == std::optional<double>(102.0));
    CHECK(book.best_ask() == std::optional<double>(103.0));
    book.validate();  // would throw on a crossed book
    return true;
}

bool sweep_stops_at_the_limit_price() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 5, 101.0), book);

    MatchResult r = engine.submit(Order(3, Side::BUY, OrderType::LIMIT, 10, 100.0), book);
    CHECK(r.trades.size() == 1);
    CHECK(r.trades[0].price == 100.0);
    CHECK(r.resting);
    CHECK(r.order.remaining == 5);
    CHECK(book.best_bid() == std::optional<double>(100.0));
    CHECK(book.best_ask() == std::optional<double>(101.0));
    book.validate();
    return true;
}

bool market_order_exhaustion_expires_the_remainder() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 5, 101.0), book);

    MatchResult r = engine.submit(Order(3, Side::BUY, OrderType::MARKET, 20), book);
    CHECK(r.trades.size() == 2);
    CHECK(r.trades[0].price == 100.0);  // best first
    CHECK(r.trades[1].price == 101.0);
    CHECK(r.order.remaining == 10);
    CHECK(!r.resting);
    CHECK(r.status() == OrderStatus::EXPIRED);
    CHECK(book.level_count(Side::SELL) == 0);
    CHECK(!book.best_ask().has_value());
    book.validate();
    return true;
}

bool market_order_on_an_empty_book_expires_immediately() {
    OrderBook book;
    MatchingEngine engine;

    MatchResult r = engine.submit(Order(1, Side::BUY, OrderType::MARKET, 10), book);
    CHECK(r.trades.empty());
    CHECK(r.order.remaining == 10);
    CHECK(!r.resting);
    CHECK(r.status() == OrderStatus::EXPIRED);
    CHECK(book.size() == 0);
    book.validate();
    return true;
}

bool limit_remainder_rests_without_crossing() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);

    MatchResult r = engine.submit(Order(2, Side::BUY, OrderType::LIMIT, 10, 100.0), book);
    CHECK(r.trades.size() == 1);
    CHECK(r.resting);
    CHECK(r.order.remaining == 5);
    CHECK(book.best_bid() == std::optional<double>(100.0));
    CHECK(!book.best_ask().has_value());
    book.validate();
    return true;
}

bool trade_ids_are_sequential_across_submissions() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);
    MatchResult r1 = engine.submit(Order(2, Side::BUY, OrderType::LIMIT, 3, 100.0), book);
    MatchResult r2 = engine.submit(Order(3, Side::BUY, OrderType::LIMIT, 2, 100.0), book);

    CHECK(r1.trades.size() == 1 && r1.trades[0].trade_id == 1);
    CHECK(r2.trades.size() == 1 && r2.trades[0].trade_id == 2);
    CHECK(engine.trades_generated() == 2);
    book.validate();
    return true;
}

bool l2_aggregates_and_caps_depth_best_first() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 3, 100.0), book);
    engine.submit(Order(3, Side::SELL, OrderType::LIMIT, 7, 101.0), book);
    engine.submit(Order(4, Side::BUY, OrderType::LIMIT, 5, 98.0), book);
    engine.submit(Order(5, Side::BUY, OrderType::LIMIT, 5, 99.0), book);

    BookSnapshot snap = book.l2();
    CHECK(snap.asks.size() == 2);
    CHECK(snap.asks[0].price == 100.0 && snap.asks[0].quantity == 8 &&
          snap.asks[0].order_count == 2);
    CHECK(snap.asks[1].price == 101.0 && snap.asks[1].quantity == 7);
    CHECK(snap.bids.size() == 2);
    CHECK(snap.bids[0].price == 99.0);  // best bid first
    CHECK(snap.bids[1].price == 98.0);

    BookSnapshot capped = book.l2(1);
    CHECK(capped.asks.size() == 1 && capped.asks[0].price == 100.0);
    CHECK(capped.bids.size() == 1 && capped.bids[0].price == 99.0);
    CHECK(snap.best_bid() == std::optional<double>(99.0));
    CHECK(snap.best_ask() == std::optional<double>(100.0));
    book.validate();
    return true;
}

bool l3_is_strict_price_time_on_both_sides() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::BUY, OrderType::LIMIT, 5, 99.0), book);
    engine.submit(Order(2, Side::BUY, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(3, Side::BUY, OrderType::LIMIT, 5, 100.0), book);
    engine.submit(Order(4, Side::SELL, OrderType::LIMIT, 5, 102.0), book);
    engine.submit(Order(5, Side::SELL, OrderType::LIMIT, 5, 101.0), book);
    engine.submit(Order(6, Side::SELL, OrderType::LIMIT, 5, 101.0), book);

    auto bids = book.l3(Side::BUY);
    CHECK(bids.size() == 3);
    CHECK(bids[0].order_id == 2 && bids[0].price == 100.0);
    CHECK(bids[1].order_id == 3 && bids[1].price == 100.0);
    CHECK(bids[2].order_id == 1 && bids[2].price == 99.0);

    auto asks = book.l3(Side::SELL);
    CHECK(asks.size() == 3);
    CHECK(asks[0].order_id == 5 && asks[0].price == 101.0);
    CHECK(asks[1].order_id == 6 && asks[1].price == 101.0);
    CHECK(asks[2].order_id == 4 && asks[2].price == 102.0);
    book.validate();
    return true;
}

bool top_of_book_spread_mid_and_depth() {
    OrderBook book;
    MatchingEngine engine;
    engine.submit(Order(1, Side::BUY, OrderType::LIMIT, 5, 99.0), book);
    engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 5, 101.0), book);

    CHECK(book.best_bid() == std::optional<double>(99.0));
    CHECK(book.best_ask() == std::optional<double>(101.0));
    CHECK(book.spread() == std::optional<double>(2.0));
    CHECK(book.mid_price() == std::optional<double>(100.0));
    CHECK(book.bid_depth() == 5 && book.ask_depth() == 5);
    CHECK(book.depth_at(Side::BUY, 99.0) == 5);
    CHECK(book.depth_at(Side::BUY, 50.0) == 0);

    engine.submit(Order(3, Side::SELL, OrderType::MARKET, 2), book);
    CHECK(book.bid_depth() == 3);
    CHECK(book.depth_at(Side::BUY, 99.0) == 3);
    book.validate();
    return true;
}

bool status_transitions_match_the_engine_contract() {
    OrderBook book;
    MatchingEngine engine;

    MatchResult rested = engine.submit(Order(1, Side::SELL, OrderType::LIMIT, 10, 100.0), book);
    CHECK(rested.status() == OrderStatus::RESTING);

    MatchResult partial = engine.submit(Order(2, Side::SELL, OrderType::LIMIT, 10, 105.0), book);
    CHECK(partial.status() == OrderStatus::RESTING);

    // Crosses partially, then rests.
    MatchResult r = engine.submit(Order(3, Side::BUY, OrderType::LIMIT, 15, 100.0), book);
    CHECK(r.status() == OrderStatus::PARTIALLY_FILLED && r.resting);

    // Fully filled: sweeps order 2 at 105 in a single trade.
    MatchResult f = engine.submit(Order(4, Side::BUY, OrderType::LIMIT, 5, 105.0), book);
    CHECK(f.status() == OrderStatus::FILLED && !f.resting);
    CHECK(f.trades.size() == 1);
    CHECK(f.trades[0].price == 105.0 && f.trades[0].resting_order_id == 2);

    book.validate();
    return true;
}

bool quantity_is_conserved_under_random_limit_flow() {
    // Limit-only flow (no expiry, no cancels): every submitted share is
    // either traded or still resting. Each trade fills both sides, so
    //   submitted == 2 * traded + resting depth.
    OrderBook book;
    MatchingEngine engine;
    std::mt19937 rng(42);

    long submitted = 0, traded = 0;
    for (int i = 0; i < 2000; ++i) {
        Side side = (rng() % 2 == 0) ? Side::BUY : Side::SELL;
        int qty = 1 + static_cast<int>(rng() % 40);
        // Bids 99.90..100.00, asks 100.00..100.10: the books cross at 100.
        double price = side == Side::BUY
                           ? 99.90 + 0.01 * static_cast<double>(rng() % 11)
                           : 100.00 + 0.01 * static_cast<double>(rng() % 11);
        MatchResult r = engine.submit(Order(i + 1, side, OrderType::LIMIT, qty, price), book);
        submitted += qty;
        for (const Trade& t : r.trades) traded += t.quantity;
    }

    CHECK(submitted == 2 * traded + book.bid_depth() + book.ask_depth());
    book.validate();
    return true;
}

}  // namespace

std::vector<testing::Case> matching_cases() {
    return {
        {"multi_level_sweep_walks_in_price_order", multi_level_sweep_walks_in_price_order},
        {"locked_book_trades_at_the_touch", locked_book_trades_at_the_touch},
        {"crossed_incoming_limit_gets_improvement_then_rests_uncrossed",
         crossed_incoming_limit_gets_improvement_then_rests_uncrossed},
        {"sweep_stops_at_the_limit_price", sweep_stops_at_the_limit_price},
        {"market_order_exhaustion_expires_the_remainder", market_order_exhaustion_expires_the_remainder},
        {"market_order_on_an_empty_book_expires_immediately", market_order_on_an_empty_book_expires_immediately},
        {"limit_remainder_rests_without_crossing", limit_remainder_rests_without_crossing},
        {"trade_ids_are_sequential_across_submissions", trade_ids_are_sequential_across_submissions},
        {"l2_aggregates_and_caps_depth_best_first", l2_aggregates_and_caps_depth_best_first},
        {"l3_is_strict_price_time_on_both_sides", l3_is_strict_price_time_on_both_sides},
        {"top_of_book_spread_mid_and_depth", top_of_book_spread_mid_and_depth},
        {"status_transitions_match_the_engine_contract", status_transitions_match_the_engine_contract},
        {"quantity_is_conserved_under_random_limit_flow", quantity_is_conserved_under_random_limit_flow},
    };
}
